internal/commands/vaultsecrets/run/run.go

```diff
@@ -0,0 +1,157 @@
+
+package run
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	preview_secret_service "github.com/hashicorp/hcp-sdk-go/clients/cloud-vault-secrets/preview/2023-11-28/client/secret_service"
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-vault-secrets/stable/2023-06-13/client/secret_service"
+	"github.com/hashicorp/hcp/internal/commands/vaultsecrets/apps/helper"
+	"github.com/hashicorp/hcp/internal/commands/vaultsecrets/secrets/appname"
+	"github.com/hashicorp/hcp/internal/pkg/cmd"
+	"github.com/hashicorp/hcp/internal/pkg/format"
+	"github.com/hashicorp/hcp/internal/pkg/heredoc"
+	"github.com/hashicorp/hcp/internal/pkg/iostreams"
+	"github.com/hashicorp/hcp/internal/pkg/profile"
+)
+
+type RunOpts struct {
+	Ctx     context.Context
+	Profile *profile.Profile
+	IO      iostreams.IOStreams
+	Output  *format.Outputter
+
+	AppName       string
+	Command       string
+	PreviewClient preview_secret_service.ClientService
+	Client        secret_service.ClientService
+}
+
+func NewCmdRun(ctx *cmd.Context, runF func(*RunOpts) error) *cmd.Command {
+	opts := &RunOpts{
+		Ctx:           ctx.ShutdownCtx,
+		Profile:       ctx.Profile,
+		IO:            ctx.IO,
+		Output:        ctx.Output,
+		PreviewClient: preview_secret_service.New(ctx.HCP, nil),
+		Client:        secret_service.New(ctx.HCP, nil),
+	}
+
+	cmd := &cmd.Command{
+		Name:      "run",
+		ShortHelp: "Run a process with secrets from a Vault Secrets app.",
+		LongHelp: heredoc.New(ctx.IO).Must(`
+		The {{ template "mdCodeOrBold" "hcp vault-secrets run" }} command lets you
+		run the provided command as a child process while injecting
+        all of the app's secrets as environment variables, with all secret names
+        converted to upper-case. The stdout/stderr from the child
+        process are forwarded to the top level {{ template "mdCodeOrBold" "hcp vault-secrets run" }} command.
+		`),
+		Examples: []cmd.Example{
+			{
+				Preamble: `Inject secrets as an environment variable:`,
+				Command: heredoc.New(ctx.IO, heredoc.WithPreserveNewlines()).Must(`
+				$ hcp vault-secrets run "env"
+				`),
+			},
+		},
+		Args: cmd.PositionalArguments{
+			Args: []cmd.PositionalArgument{
+				{
+					Name:          "COMMAND",
+					Documentation: "Defines the invocation of the child process to inject secrets to.",
+				},
+			},
+		},
+		Flags: cmd.Flags{
+			Local: []*cmd.Flag{
+				{
+					Name:         "app",
+					DisplayValue: "NAME",
+					Description:  "The application you want to pull all secrets from.",
+					Value:        appname.Flag(),
+				},
+			},
+		},
+		PersistentPreRun: func(c *cmd.Command, args []string) error {
+			return appname.Require(ctx)
+		},
+		RunF: func(c *cmd.Command, args []string) error {
+			opts.Command = args[0]
+
+			opts.AppName = appname.Get()
+
+			if runF != nil {
+				return runF(opts)
+			}
+			return runRun(opts)
+		},
+	}
+	cmd.Args.Autocomplete = helper.PredictAppName(ctx, cmd, preview_secret_service.New(ctx.HCP, nil))
```

Comment:
can use opts.PreviewClient

Reply:
cmd.Args.Autocomplete is for COMMAND arg, but auto complete is for app flag name. Is this correct usage?

Reply:
Yeah you'll want to set the flag `app` autocomplete not the args